### PR TITLE
Update to InteractableToggleCollection - Fix for 6274

### PR DESCRIPTION
**Pull request by CDiaz-MS**
## Overview

Fix for issue #6274 

Changing the CurrentIndex property in InteractableToggleCollection via script or the inspector did not update toggle states within the group.  

The changes added get/set methods to the CurrentIndex property and updated SetSelection to trigger an OnClick event. 

## InteractableToggleCollection Behavior BEFORE 
BeforeITCRefactor

## InteractableToggleCollection Behavior AFTER 
AfterITCRefactor

## Changes
- Added custom inspector for InteractableToggleCollection
- Updated InteractableToggleCollection.cs
- Added Test for the changes in InteractableTests

## Verification
- pull the changes
- open unity project
- add MRTK to the unity project
- add the RadialSet prefab
- press play and change the index in the InteractableToggleCollection, the toggle should be updated to match the index input

> As a reviewer, it is possible to check out this change locally by using the following
> commands (substituting {PR_ID} with the ID of this pull request):
>
> git fetch origin pull/{PR_ID}/head:name_of_local_branch
>
> git checkout name_of_local_branch

### CODE REVIEW

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs [hunk 1]
```diff
@@ -11,78 +12,108 @@ namespace Microsoft.MixedReality.Toolkit.UI
     /// </summary>
     public class InteractableToggleCollection : MonoBehaviour
     {
-        [Tooltip("Interactables that will be managed by this controller")]
+        [Tooltip("Array of Interactables that will be managed by this controller")]
+        /// <summary>
+        /// Array of Interactables that will be managed by this controller
+        /// </summary>
         public Interactable[] ToggleList;
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: ToggleList does not have get/set and backing serialized private field etc

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs [hunk 2]
```diff
@@ -11,78 +12,108 @@ namespace Microsoft.MixedReality.Toolkit.UI
...
+        private int currentIndex;
 
-        private void Start()
+        /// <summary>
+        /// The current index in the array of interactable toggles
+        /// </summary>
+        public int CurrentIndex
         {
-            for (int i = 0; i < ToggleList.Length; ++i)
+            get
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: simplify to ```=> currentIndex```

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs [hunk 3]
```diff
@@ -11,78 +12,108 @@ namespace Microsoft.MixedReality.Toolkit.UI
...
+        private void Start()
         {
-            OnSelection(CurrentIndex, true);
+            if (ToggleList != null)
+            {
+                // Add listeners to each toggle in ToggleList
+                for (int i = 0; i < ToggleList.Length; ++i)
+                {
+                    int itemIndex = i;
+                    ToggleList[i].OnClick.AddListener(() => SetSelection(itemIndex, true, false));
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: You need to save list of actions and then remove this list. The removelistener later is a different anonymous instance

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs [hunk 4]
```diff
@@ -11,78 +12,108 @@ namespace Microsoft.MixedReality.Toolkit.UI
...
+
+                SetSelection(CurrentIndex, true, true);
+            }
         }
 
         /// <summary>
-        /// Sets the selected index and selected Interactive
+        /// Set the selection of a an element in the toggle collection based on index.
+        /// <param name="index">Index of an element in ToggleList</param>
+        /// <param name="force">Force update currentIndex</param>
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: nit: I would update this to "force selection set"

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs [hunk 5]
```diff
@@ -11,78 +12,108 @@ namespace Microsoft.MixedReality.Toolkit.UI
...
+        /// <param name="fireOnClick">The manual trigger of the OnClick event. OnClick event is manually triggered 
+        /// when the CurrentIndex is updated via script or inspector</param>
         /// </summary>
-        public void SetSelection(int index)
+        public void SetSelection(int index, bool force = false, bool fireOnClick = false)
         {
-            if (!isActiveAndEnabled ||
-                (index < 0 || ToggleList.Length <= index))
+            if (index < 0 || ToggleList.Length <= index)
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: case if toggellist is null?

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs [hunk 6]
```diff
@@ -11,78 +12,108 @@ namespace Microsoft.MixedReality.Toolkit.UI
...
-        /// </summary>
+        // Update the visual appearance and set the states of the selected and unselected toggles within 
+        // Interactable
         protected virtual void OnSelection(int index, bool force = false)
         {
-            for (int i = 0; i < ToggleList.Length; ++i)
+            int length = ToggleList.Length;
+            for (int i = 0; i < length; ++i)
             {
                 if (i != index)
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: simplify 
ToggleList[I].IsToggled = i == index;

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs [hunk 7]
```diff
@@ -11,78 +12,108 @@ namespace Microsoft.MixedReality.Toolkit.UI
...
+            OnSelectionEvents.Invoke();
         }
 
         private void OnDestroy()
         {
             for (int i = 0; i < ToggleList.Length; ++i)
             {
                 int itemIndex = i;
-                ToggleList[i].OnClick.RemoveListener(() => OnSelection(itemIndex));
+                ToggleList[i].OnClick.RemoveListener(() => SetSelection(itemIndex, true, false));
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: might be safe to check ToggleList[i] is not null

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs [hunk 8]
```diff
@@ -11,78 +12,108 @@ namespace Microsoft.MixedReality.Toolkit.UI
...
-
-            if (force)
-            {
-                ToggleList[index].IsToggled = true;
-            }
-            else
-            {
-                OnSelectionEvents.Invoke();
-            }
+            OnSelectionEvents.Invoke();
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: We actually should start always doing
OnSelectionEvents?.Invoke();

because OnSelectionEvents is public and thus could be set to null by another party

File: Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+                        else
+                        {
+                            Debug.LogError("Index out of range");
+                        }
+
+                    }
+                }
+            }
+        }
+
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs]: nit: fix spacing

File: Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using UnityEditor;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.UI.Editor
+{
+    [CustomEditor(typeof(InteractableToggleCollection))]
+    public class InteractableToggleCollectionInspector : UnityEditor.Editor
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs]: @julenka, what do you want to do with inspector public class comments? Mark internal or just add generic comment "I am inspector of blah"

File: Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        protected virtual void OnEnable()
+        {
+            instance = (InteractableToggleCollection)target;
+        }
+
+        public override void OnInspectorGUI()
+        {
+            base.OnInspectorGUI();
+
+            if (Application.isPlaying)
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs]: combine if statement conditions into one if

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs [hunk 12]
```diff
@@ -11,78 +12,108 @@ namespace Microsoft.MixedReality.Toolkit.UI
...
 
-            OnSelection(index, true);
+            if (CurrentIndex != index || force)
+            {
+                currentIndex = index;
+
+                OnSelection(index);
+            }
+
+            if (fireOnClick)
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: I believe this needs to be inside the if statement body above. If we aren't changing selection, we shouldn't be firing a click? @julenka thoughts?

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs [hunk 13]
```diff
@@ -11,78 +12,108 @@ namespace Microsoft.MixedReality.Toolkit.UI
...
-            for (int i = 0; i < ToggleList.Length; ++i)
+            get
             {
-                int itemIndex = i;
-                // add selection event handler to each button
-                ToggleList[i].OnClick.AddListener(() => OnSelection(itemIndex));
-                ToggleList[i].CanDeselect = false;
+                return currentIndex;
+            }
+            set
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: simplify set => SetSelection(value, false, true);

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs (hunk 12 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: Like if I call CurrentIndex =2, many times over, seems weird to call onclick?

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/InteractableTests.cs [hunk 14]
```diff
@@ -627,6 +629,58 @@ public IEnumerator TestPressableToggleHoloLens2()
...
+                out interactableToggleCollection,
+                numRadials,
+                Vector3.forward);
+
+            var toggleList = interactableToggleCollection.ToggleList;
+
+            int[] onClickEventCalled = new int[numRadials];
+
+            int index = 0;
+            foreach (var radial in toggleList)
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/InteractableTests.cs]: nit: I'd suggest simplifying this to for (int index = 0; blah blah), foreach purpose is defeated here

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/InteractableTests.cs [hunk 15]
```diff
@@ -627,6 +629,58 @@ public IEnumerator TestPressableToggleHoloLens2()
...
+            {
+                interactableToggleCollection.CurrentIndex = j;
+                yield return null;
+
+                // If the CurrentIndex is changed the toggle should be visually updated and events should be triggered
+                for (int i = 0; i < numRadials; i++)
+                {
+                    if (i == interactableToggleCollection.CurrentIndex)
+                    {
+                        Assert.True(interactableToggleCollection.ToggleList[i].IsToggled);
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/InteractableTests.cs]: simplify to 

assert.areequals(interactableToggleCollection.ToggleList[i].IsToggled, i == interactableToggleCollection.CurrentIndex)
or
bool shouldBeSelected = i == interactableToggleCollection.CurrentIndex;
Assert.AreEquals(shouldBeSelected, interactableToggleCollection.ToggleList[i].IsToggled)

for the second assert
bool expectedClickCount = i <= j ? 1 : 0;
Assert.AreEquals(onClickEventCalled[i], expectedClickCount);

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/InteractableTests.cs [hunk 16]
```diff
@@ -33,6 +33,8 @@ public class InteractableTests : BasePlayModeTests
         private const string DefaultInteractablePrefabAssetPath = "Assets/MixedRealityToolkit.Examples/Demos/UX/Interactables/Prefabs/Model_PushButton.prefab";
         private const string RadialSetPrefabAssetPath = "Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Prefabs/RadialSet.prefab";
         private const string PressableHoloLens2TogglePrefabPath = "Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Prefabs/PressableButtonHoloLens2Toggle.prefab";
+        private const string RadialPrefabAssetPath = "Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Prefabs/Radial.prefab";
+        private const string ToggleSwitchPrefabAssetPath = "Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Prefabs/ToggleSwitch.prefab";
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/InteractableTests.cs]: use?

File: Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+            {
+                if (instance != null)
+                {
+                    if (GUI.changed)
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs]: actually don't believe this will work. If I set CurrentIndex = -1 in the inspector, it will actually set the value.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs (hunk 13 above)
Comment (CDiaz-MS) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: Simplified

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs (hunk 4 above)
Comment (CDiaz-MS) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: Updated

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs (hunk 5 above)
Comment (CDiaz-MS) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: Added a null check

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs (hunk 8 above)
Comment (CDiaz-MS) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: fixed

File: Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs (hunk 11 above)
Comment (CDiaz-MS) [reply Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs]: Combined

File: Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs (hunk 17 above)
Comment (CDiaz-MS) [reply Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs]: Added if index < 0

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/InteractableTests.cs (hunk 16 above)
Comment (CDiaz-MS) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/InteractableTests.cs]: The `RadialPrefabAssetPath` is used in `AssembleRadialSet()`.  The `ToggleSwitchPrefabAssetPath` was used to test if the ToggleSwitch prefab worked with the new InteractableToggleCollection (it does work), but it was there for testing and is not used anywhere else, so it will be removed.

Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/InteractableTests.cs]: Sorry I just meant ToggleSwitchPrefabAssetPath

File: Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs (hunk 10 above)
Comment (julenka) [reply Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs]: I like marking them as internal

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs (hunk 12 above)
Comment (julenka) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: I think Catherine had a specific case she was fixing which required onclick to get called, I think otherwise the visuals didn't update? I don't fully remember...

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs [hunk 18]
```diff
@@ -11,78 +12,108 @@ namespace Microsoft.MixedReality.Toolkit.UI
...
 
-        /// <summary>
-        /// Set the toggle state of each button based on the selected item
-        /// </summary>
+        // Update the visual appearance and set the states of the selected and unselected toggles within 
+        // Interactable
         protected virtual void OnSelection(int index, bool force = false)
         {
-            for (int i = 0; i < ToggleList.Length; ++i)
+            int length = ToggleList.Length;
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: super minor, but you don't need to pull this out into a separate variable lengths, since length is just used once. Just stylistically to be consistent with other code.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs [hunk 19]
```diff
@@ -11,78 +12,108 @@ namespace Microsoft.MixedReality.Toolkit.UI
...
+        /// Set the selection of a an element in the toggle collection based on index.
+        /// <param name="index">Index of an element in ToggleList</param>
+        /// <param name="force">Force update currentIndex</param>
+        /// <param name="fireOnClick">The manual trigger of the OnClick event. OnClick event is manually triggered 
+        /// when the CurrentIndex is updated via script or inspector</param>
         /// </summary>
-        public void SetSelection(int index)
+        public void SetSelection(int index, bool force = false, bool fireOnClick = false)
         {
-            if (!isActiveAndEnabled ||
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: Why did you need to remove the isActiveAndEnabled check here?

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs (hunk 18 above)
Comment (CDiaz-MS) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: Accidentally added the wrong reaction, meant to do thumbs up

File: Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs (hunk 10 above)
Comment (CDiaz-MS) [reply Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs]: @Troy-Ferrell and @julenka How do I mark something as internal?

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs (hunk 6 above)
Comment (CDiaz-MS) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: Simplified

File: Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs (hunk 10 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableToggleCollectionInspector.cs]: `internal class InteractableToggleCollectionInspector `

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/InteractableTests.cs (hunk 15 above)
Comment (CDiaz-MS) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/InteractableTests.cs]: Simplified

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs (hunk 1 above)
Comment (CDiaz-MS) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: Added to ToggleList:
```
        [Tooltip("Array of Interactables that will be managed by this controller")]
        [SerializeField, FormerlySerializedAs("ToggleList")]
        private Interactable[] toggleList;

        /// <summary>
        /// Array of Interactables that will be managed by this controller
        /// </summary>
        public Interactable[] ToggleList
        {
            get => toggleList;
            set => toggleList = value;
        }
```

Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: You need to update the selection on the set. If I change the size of the list, then my current index may be invalid

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs (hunk 19 above)
Comment (CDiaz-MS) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: The removal of that check was not intentional, it may have been accidentally removed earlier.  Added it back.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs (hunk 12 above)
Comment (CDiaz-MS) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: There was no specific case.  The fireOnClick was moved into the index changed and force check.  After the move, if I changed the index from the inspector at runtime the toggle collection was not updated.  Setting force to true in the Inspector script solved this issue.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs (hunk 1 above)
Comment (CDiaz-MS) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/InteractableToggleCollection.cs]: Addressed in PR #6523
